usb: hid: allow boot interface Protocol Code to be set per device

jfischer-no opened the pull request:
Kconfig option USB_HID_PROTOCOL_CODE does not allow to set
boot interface protocol code for specific HID device but
only to set the same value for all device.
Add new API function to allow the application to set
Protocol Code per device. Deprecate USB_HID_PROTOCOL_CODE option.

Fixes: #32778

jfischer-no committed:
usb: hid: allow boot interface Protocol Code to be set per device

Kconfig option USB_HID_PROTOCOL_CODE does not allow to set
boot interface protocol code for specific HID device but
only to set the same value for all device.
Add new API function to allow the application to set
Protocol Code per device. Deprecate USB_HID_PROTOCOL_CODE option.

Fixes: #32778

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: hid: set boot interface Protocol Code

Update sampele to show how to set boot interface Protocol Code.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on subsys/usb/class/hid/core.c:
```diff
@@ -675,6 +675,19 @@ static void hid_interface_config(struct usb_desc_header *head,
+	struct usb_cfg_data *cfg = (void *)dev->config;
+	struct usb_if_descriptor *if_desc = (void *)cfg->interface_descriptor;
+
+	if (IS_ENABLED(CONFIG_USB_HID_BOOT_PROTOCOL)) {
```
I think no, reserved may be still valid.

jfischer-no commented on subsys/usb/class/hid/Kconfig:
```diff
@@ -44,12 +44,18 @@ config USB_HID_POLL_INTERVAL_MS
 	  Polling interval in ms selected by the USB HID Device.
 
 config USB_HID_REPORTS
```
I guess yes, but how I see it just user experience there is no benefit from this change in the code. And that needs to be done in #32355 :-)

MarekPieta commented on subsys/usb/class/hid/core.c:
```diff
@@ -675,6 +675,19 @@ static void hid_interface_config(struct usb_desc_header *head,
+int usb_hid_set_proto_code(const struct device *dev, uint8_t proto_code)
+{
+	struct usb_cfg_data *cfg = (void *)dev->config;
+	struct usb_if_descriptor *if_desc = (void *)cfg->interface_descriptor;
```
Do we need the explicit casts here ?

jfischer-no replied:
ohh, copy & paste from above, but `void *interface_descriptor;` does not need to have const in struct usb_cfg_data().

jfischer-no replied:
fixed it for this patch.

jfischer-no committed:
usb: remove const qualifier in struct usb_cfg_data()

The pointer to interface descriptor in struct usb_cfg_data
is used to access and modify the descriptor at runtime.
Remove const qualifier and thus avoid unnecessary casts.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
